prow/cmd/exporter/collector.go

```diff
@@ -169,3 +171,153 @@ func sanitizeLabelName(s string) string {
+		pjc.lastKnownState[newJob.UID] = lastState{
+			state:    newJob.Status.State,
+			observed: now,
+		}
+		return
+	} else if lastKnownState.state != newJob.Status.State {
+		pjc.lastKnownState[newJob.UID] = lastState{
+			state:    newJob.Status.State,
+			observed: now,
+		}
+
+		labels := getJobLabel(lastKnownState, newJob)
+		histogram, err := pjc.HistogramVec.GetMetricWithLabelValues(labels.values()...)
+		if err != nil {
+			logrus.WithError(err).Error("Failed to get a histogram for a prowjob")
+			return
+		}
+
+		histogram.Observe(now.Sub(lastKnownState.observed).Seconds())
+	}
+}
+
+func (pjc *prowJobLifecycleCollector) Delete(job *prowapi.ProwJob) {
+	if _, exists := pjc.lastKnownState[job.UID]; exists {
+		delete(pjc.lastKnownState, job.UID)
+	}
+}
+
+// NewProwJobLifecycleCollector creates histograms which can track the timespan between ProwJob state transitions.
+// The histograms are based on the job name, the old job state and the new job state.
+// Data is collected by hooking itself into the prowjob informer.
+// The collector will never record the same state transition twice, even if reboots happen.
+// The collector will not record the first state transition it sees after booting, and therefore miss the first transition of each job after a restart.
+// TODO: by looking at the created pods we can collect more transitions on restarts
+func NewProwJobLifecycleCollector(informer cache.SharedIndexInformer) *prowJobLifecycleCollector {
+	lifecycleCollector := &prowJobLifecycleCollector{
+		lastKnownState: map[types.UID]lastState{},
+		HistogramVec: prometheus.NewHistogramVec(
+			prometheus.HistogramOpts{
+				Name: "prow_job_runtime_seconds",
+				Buckets: []float64{
+					time.Minute.Seconds() / 2,
+					(1 * time.Minute).Seconds(),
+					(2 * time.Minute).Seconds(),
+					(5 * time.Minute).Seconds(),
+					(10 * time.Minute).Seconds(),
+					(1 * time.Hour).Seconds() / 2,
+					(1 * time.Hour).Seconds(),
+					(2 * time.Hour).Seconds(),
+					(3 * time.Hour).Seconds(),
+					(4 * time.Hour).Seconds(),
+					(5 * time.Hour).Seconds(),
+					(6 * time.Hour).Seconds(),
+					(7 * time.Hour).Seconds(),
+					(8 * time.Hour).Seconds(),
+					(9 * time.Hour).Seconds(),
+					(10 * time.Hour).Seconds(),
+				},
+			},
+			[]string{
+				// namespace of the job
+				"job_namespace",
+				// name of the job
+				"job_name",
+				// type of the prowjob: presubmit, postsubmit, periodic, batch
+				"type",
+				// last state of the prowjob: triggered, pending, success, failure, aborted, error
+				"last_state",
+				// state of the prowjob: triggered, pending, success, failure, aborted, error
+				"state",
+				// the org of the prowjob's repo
+				"org",
+				// the prowjob's repo
+				"repo",
+				// the base_ref of the prowjob's repo
+				"base_ref",
+			},
+		),
+	}
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(job interface{}) {
+			lifecycleCollector.Update(nil, job.(*prowapi.ProwJob))
```

Comment:
Not sure if this is an issue, but the current architecture means that we process completely synchronously: https://github.com/kubernetes/kubernetes/blob/2d8b8703de920bafaf0b893582e18e5bdd03c51d/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L652

Has the advantage that we don't need locking on the map thought.

Reply:
I think that the operations in `Update` are lightweight enough to do it here already, and in contrast  to controllers we need to catch the actual transitions. If we think it is too heavy-weight, I could push the jobs into a channel with a buffer.